File: backend/kotlin/backend-kotlin-spring/src/main/kotlin/dev/reference/app/service/AuthService.kt
package dev.reference.app.service

import dev.reference.app.dto.AuthResponse
import dev.reference.app.dto.LoginRequest
import dev.reference.app.dto.RegisterRequest
import dev.reference.app.dto.UserProfileResponse
import dev.reference.app.entity.UserEntity
import dev.reference.app.exception.AuthException
import dev.reference.app.repository.UserRepository
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class AuthService(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder,
    private val jwtService: JwtService,
) {
    @Transactional
    fun register(request: RegisterRequest): AuthResponse {
        if (userRepository.existsByUsername(request.username)) {
            throw AuthException(409, "Username already taken")
        }

        val user = UserEntity(
            username = request.username,
            passwordHash = passwordEncoder.encode(request.password),
        )
        try {
            // Flush inside the try so a concurrent insert that won the race surfaces here as 409
            // rather than escaping as a commit-time 500.
            userRepository.saveAndFlush(user)
        } catch (ex: DataIntegrityViolationException) {
            throw AuthException(409, "Username already taken")
        }
        return buildAuthResponse(user.username)
    }

    fun login(request: LoginRequest): AuthResponse {
        val user = userRepository.findByUsername(request.username)
            .orElseThrow { AuthException(401, "Wrong login or password") }

        if (!passwordEncoder.matches(request.password, user.passwordHash)) {
            throw AuthException(401, "Wrong login or password")
        }

        return buildAuthResponse(user.username)
    }

    fun profile(username: String): UserProfileResponse {
        userRepository.findByUsername(username)
            .orElseThrow { AuthException(401, "Unauthorized") }
        return UserProfileResponse(username)
    }

    /**
     * Authenticated self-delete. Tokens are stateless, so a JWT issued earlier keeps verifying
     * after deletion — but every endpoint that resolves the user ([profile], this one) answers
     * 401 once the row is gone. Also lets test suites clean up the users they register.
     */
    @Transactional
    fun deleteAccount(username: String) {
        val user = userRepository.findByUsername(username)
            .orElseThrow { AuthException(401, "Unauthorized") }
        userRepository.delete(user)
    }

    private fun buildAuthResponse(username: String): AuthResponse =
        AuthResponse(
            jwtService.createToken(username),
            username,
            POST_AUTH_REDIRECT,
        )

    companion object {
        private const val POST_AUTH_REDIRECT = "/"
    }
}

File: backend/javascript/backend-javascript-nest/src/auth.service.js
'use strict';

const { Injectable } = require('@nestjs/common');

const { ApiException, DuplicateUsernameError } = require('./errors');
const { hashPassword, checkPassword } = require('./passwords');
const { validateCredentials, isJsonObject } = require('./validation');
const { injectConstructor } = require('./di');
const { JwtService } = require('./jwt.service');
const { STORE, SETTINGS } = require('./tokens');

@Injectable()
class AuthService {
  constructor(store, jwt, settings) {
    this.store = store;
    this.jwt = jwt;
    this.postAuthRedirect = settings.postAuthRedirect;
  }

  authResponse(username) {
    return {
      token: this.jwt.createToken(username),
      username,
      redirectUrl: this.postAuthRedirect,
    };
  }

  /**
   * A body that was not a JSON object reaches here as null from lenientJson().
   * A parsed `{}` is not that case: it goes through field validation.
   */
  credentialsOrThrow(body) {
    if (!isJsonObject(body)) {
      throw new ApiException(400, 'Request body is not valid JSON');
    }
    const error = validateCredentials(body.username, body.password);
    if (error) {
      throw new ApiException(400, error);
    }
    return { username: body.username, password: body.password };
  }

  async register(body) {
    const { username, password } = this.credentialsOrThrow(body);
    if (await this.store.findUserByUsername(username)) {
      throw new ApiException(409, 'Username already taken');
    }
    try {
      await this.store.insertUser(username, hashPassword(password));
    } catch (error) {
      // A concurrent insert that won the race surfaces as the unique violation.
      if (error instanceof DuplicateUsernameError) {
        throw new ApiException(409, 'Username already taken');
      }
      throw error;
    }
    return this.authResponse(username);
  }

  async login(body) {
    const { username, password } = this.credentialsOrThrow(body);
    const user = await this.store.findUserByUsername(username);
    if (user === null || !checkPassword(password, user.passwordHash)) {
      throw new ApiException(401, 'Wrong login or password');
    }
    return this.authResponse(username);
  }

  /** Resolves the bearer token to a username that still exists, or throws 401. */
  async profile(authorizationHeader) {
    const header =
      typeof authorizationHeader === 'string' ? authorizationHeader : '';
    if (!header.startsWith('Bearer ')) {
      throw new ApiException(401, 'Unauthorized');
    }
    const username = this.jwt.extractUsername(header.slice('Bearer '.length));
    if (!username) {
      throw new ApiException(401, 'Unauthorized');
    }
    if ((await this.store.findUserByUsername(username)) === null) {
      throw new ApiException(401, 'Unauthorized');
    }
    return { username };
  }

  /**
   * Authenticated self-delete. Tokens are stateless, so a JWT issued earlier keeps
   * verifying after deletion — but AuthGuard answers 401 once the row is gone.
   */
  async deleteAccount(username) {
    await this.store.deleteUser(username);
  }
}

injectConstructor(AuthService, STORE, JwtService, SETTINGS);

module.exports = { AuthService };
